### ISSUE TITLE
Feedforward, self-attention and layernorm computational order in StackedSelfAttentionEncoder.

### ISSUE THREAD

**Issue by yongtaoge** [allenai/allennlp#3334]
**Question**
In the original paper, the computational order may like this:
```
x -> attention(x) -> x + dropout(attention(x)) -> layernorm(x + dropout(attention(x))) => y
y -> feedforward(y) -> y + dropout(feedforward(y)) -> layernorm(y + dropout(feedforward(y))) 
=> z(input of next layer)
```
but I found the compuational order in StackedSelfAttentionEncoder is:
```
x -> feedforward(x) -> dropout(feedforward(x)) -> layernorm(x + dropout(feedforward(x))) => y
y -> attention(y) -> dropout(attention(y)) -> y + dropout(attention(y)) -> layernorm(y + dropout(attention(y)))
=> z(input of next layer)
```
The FFN is computed after multi-head attention in the original paper, but in allennlp implenmentation, the FFN is computed before multi-head attention. Maybe I miss something?
https://github.com/allenai/allennlp/blob/200239712738815a49c24f86961762b6f1636513/allennlp/modules/seq2seq_encoders/stacked_self_attention.py#L156-L164

Do I have a wrong understanding of the code?

**Comment by matt-gardner** [comment]
Those look identical to me. And can you give links to the paper and the code you're talking about? It's hard to answer the question with so little to go off of.

**Comment by yongtaoge** [comment]
![attention](https://user-images.githubusercontent.com/22744013/66450212-44131b00-ea8a-11e9-999f-19bc0f9b128c.png)
The FFN is computed after multi-head attention in the original paper, but in allennlp implenmentation, the FFN is computed before multi-head attention. Maybe I miss something?
https://github.com/allenai/allennlp/blob/200239712738815a49c24f86961762b6f1636513/allennlp/modules/seq2seq_encoders/stacked_self_attention.py#L156-L164

**Comment by yongtaoge** [comment]
Also, the tensor2tensor suggests that learning is more robust when preprocessing each layer with layernorm and postprocessing with: `dropout -> add residual` instead of postprocessing each layer with layernorm. I found fairseq support this feature in:
https://github.com/pytorch/fairseq/blob/master/fairseq/modules/transformer_layer.py

**Comment by DeNeutoy** [comment]
Hi @yongtaoge 

from what I remember when I implemented this, it is because I was following this:

https://github.com/XMUNLP/Tagger/blob/master/models/deepatt.py#L123

because I was trying to use it to implement this paper:

https://arxiv.org/abs/1712.01586

and this was before there was a "canonical" formulation of what the internals of a transformer should look like. Basically, we should probably use the `MultiheadedSelfAttention` layer in pytorch, as this is likely to stay more static/truthful. We actually have several transformer implementations in the repository currently haha, see #2270 .

So in summary, there's no particular reason it's like that, and i'd recommend that you use the pytorch version. It would be a bit of a pain to change this now, as if anyone had used it, it would silently break. So I think i'll close this - hopefully that explanation was sufficient!
